plugins/woocommerce/src/Blocks/BlockTypes/ProductCollection.php

```diff
@@ -76,13 +84,115 @@ protected function initialize() {
 
+		// Provide location context into block's context.
+		add_filter( 'render_block_context', array( $this, 'update_context' ), 11, 3 );
+
 		// Interactivity API: Add navigation directives to the product collection block.
 		add_filter( 'render_block_woocommerce/product-collection', array( $this, 'enhance_product_collection_with_interactivity' ), 10, 2 );
 		add_filter( 'render_block_core/query-pagination', array( $this, 'add_navigation_link_directives' ), 10, 3 );
 
 		add_filter( 'posts_clauses', array( $this, 'add_price_range_filter_posts_clauses' ), 10, 2 );
 	}
 
+	/**
+	 * Provides the location context to each inner block of the product collection block.
+	 *
+	 * The context schema is an array with the following data:
+	 * - type: The context type. Possible values are 'site', 'order', 'cart', 'archive', 'product'.
+	 * - sourceData: The context data.
+	 *
+	 * The sourceData structure depends on the context type as follows:
+	 * - site: []
+	 * - order: ['orderId' => int]
+	 * - cart: ['productIds' => array]
+	 * - archive: ['taxonomy' => string, 'termId' => int]
+	 * - product: ['productId' => int]
+	 *
+	 * @param array    $context      The block context.
+	 * @param array    $block        The parsed block.
+	 * @param WP_Block $parent_block The parent block.
+	 *
+	 * @return array The block's context including the location context.
+	 */
+	public function update_context( $context, $block, $parent_block ) {
+		// Run only on frontend.
+		// This is needed to avoid SSR renders within patterns. @see https://github.com/woocommerce/woocommerce/issues/45181.
+		if ( is_admin() || \WC()->is_rest_api_request() ) {
+			return $context;
+		}
+
+		if ( 'woocommerce/product-collection' === $block['blockName']
+			&& isset( $block['attrs']['query']['isProductCollectionBlock'] ) ) {
+					$this->product_collection_inner_blocks_names = array_reverse(
+						$this->extract_inner_block_names( $block )
+					);
+		}
+
+		$this->provide_location_context_inner_blocks( $block, $context );
+
+		return $context;
+	}
+
+	/**
+	 * Extract the inner block names for the Product Collection block.
+	 *
+	 * @param array $block The Product Collection block.
+	 * @param array $result Array of inner block names.
+	 *
+	 * @return array Array containing all the inner block names.
+	 */
+	protected function extract_inner_block_names( $block, &$result = array() ) {
+		if ( isset( $block['blockName'] ) ) {
+			$result[] = $block['blockName'];
+		}
+
+		if ( isset( $block['innerBlocks'] ) ) {
+			foreach ( $block['innerBlocks'] as $inner_block ) {
+				$this->extract_inner_block_names( $inner_block, $result );
+			}
+		}
+		return $result;
+	}
+
+	/**
+	 * Add the productCollectionLocation context data into the inner blocks' context.
+	 *
+	 * @param array $block Block attributes.
+	 * @param array $context Block context.
+	 */
+	protected function provide_location_context_inner_blocks( $block, &$context ) {
```

Comment:
Yup! I wouldn't say this is about best practices but rather about following a constistent way of handling inner block within the codebase.

Some points to consider:
- PHP only passes _objects_ by reference; arrays are not objects, so you get a copy of them within the method's arguments. More details can be found here: [array](http://php.net/manual/en/language.types.array.php), which states (quoting) :
```
Array assignment always involves value copying. Use the reference operator to copy an array by reference.
```
- The use of the reference operator in this case can certainly help with performance (but slightly since the array will be quite small.) 

So, in general, I think this works as expected.

Reply:
TIL, Thank you for explaining this to me 🙌🏻